modules/examples/sources/threaded-random-generator/threaded-random-generator.c

```diff
@@ -0,0 +1,185 @@
+
+          LogMessage *msg = log_msg_new_empty();
+          log_msg_set_value(msg, LM_V_MESSAGE, random_hex_str, -1);
+
+          log_threaded_source_blocking_post(s, msg);
+        }
+
+      usleep(self->freq * 1000);
+    }
+
+  g_free(random_hex_str);
+  g_free(random_bytes);
+}
+
+static void
+_request_exit(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  g_atomic_counter_set(&self->exit_requested, TRUE);
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  if (!self->bytes)
+    {
+      msg_error("The bytes() option for random-generator() is mandatory", log_pipe_location_tag(s));
+      return FALSE;
+    }
+
+  return log_threaded_source_driver_init_method(s);
+}
+
+static const gchar *
+_format_stats_instance(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+  static gchar persist_name[1024];
+
+  if (s->super.super.super.persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "random-generator,%s", s->super.super.super.persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "random-generator,%u,%u,%d", self->bytes, self->freq, self->urandom);
+
+  return persist_name;
+}
+
+LogSourceOptions *
+threaded_random_generator_sd_get_source_options(LogDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  return &self->super.worker_options.super;
+}
+
+void
+threaded_random_generator_sd_set_freq(LogDriver *s, guint freq)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+  self->freq = freq;
+}
+
+void
+threaded_random_generator_sd_set_bytes(LogDriver *s, guint bytes)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+  self->bytes = bytes;
+}
+
+gboolean
+threaded_random_generator_sd_set_type(LogDriver *s, const gchar *type)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  if (strcasecmp(type, "random") == 0)
+    self->urandom = FALSE;
+  else if (strcasecmp(type, "urandom") == 0)
+    self->urandom = TRUE;
+  else
+    return FALSE;
+
+  return TRUE;
+}
+
+LogDriver *
+threaded_random_generator_sd_new(GlobalConfig *cfg)
+{
+  ThreadedRandomGeneratorSourceDriver *self = g_new0(ThreadedRandomGeneratorSourceDriver, 1);
+  log_threaded_source_driver_init_instance(&self->super, cfg);
+
+  self->freq = 1000;
+  g_atomic_counter_set(&self->exit_requested, FALSE);
+
+  log_threaded_source_driver_set_worker_run(&self->super, _run);
+  log_threaded_source_driver_set_worker_request_exit(&self->super, _request_exit);
+
+  self->super.super.super.super.init = _init;
+  self->super.format_stats_instance = _format_stats_instance;
+
```

Comment:
hmm.. the how is the stats source specified?

Reply:
What do you mean? This name is used by the `LogSource` instance inside `LogThreadedSourceDriver` to track stats (`recvd_messages`, for example).